core/src/test/java/org/apache/iceberg/TestSchemaUpdateSync.java

```diff
@@ -0,0 +1,300 @@
+  }
+
+  @Test
+  // int	32-bit signed integers -> Can promote to long
+  public void testTypePromoteIntegerToLong() {
+    Schema currentSchema = new Schema(required(1, "aCol", IntegerType.get()));
+    Schema newSchema = new Schema(required(1, "aCol", LongType.get()));
+
+    Schema applied = new SchemaUpdate(currentSchema, 1).upsertSchema(newSchema).apply();
+    Assert.assertEquals(1, applied.asStruct().fields().size());
+    Assert.assertEquals(LongType.get(), applied.asStruct().fields().get(0).type());
+  }
+
+  @Test
+  // float	32-bit IEEE 754 floating point -> Can promote to double
+  public void testTypePromoteFloatToDouble() {
+    Schema currentSchema = new Schema(required(1, "aCol", FloatType.get()));
+    Schema newSchema = new Schema(required(1, "aCol", DoubleType.get()));
+
+    Schema applied = new SchemaUpdate(currentSchema, 1).upsertSchema(newSchema).apply();
+    Assert.assertEquals(1, applied.asStruct().fields().size());
+    Assert.assertEquals(DoubleType.get(), applied.asStruct().fields().get(0).type());
+    // Doesn't work Assert.assertEquals(newSchema.asStruct(), applied.asStruct());
+    // java.lang.AssertionError:
+    // Expected :struct<1: aCol: required double>
+    // Actual   :struct<1: aCol: required double ()>
+  }
+
+  @Test
+  public void testInvalidTypePromoteDoubleToFloat() {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Cannot change column type: aCol: double -> float");
+
+    Schema currentSchema = new Schema(required(1, "aCol", DoubleType.get()));
+    Schema newSchema = new Schema(required(1, "aCol", FloatType.get()));
+
+    new SchemaUpdate(currentSchema, 1).upsertSchema(newSchema).apply();
+  }
+
+  @Test
+  // decimal(P,S)	Fixed-point decimal; precision P, scale S	-> Scale is fixed [1], precision must be 38 or less
+  public void testTypePromoteDecimalToFixedScaleWithWiderPrecision() {
+    Schema currentSchema = new Schema(required(1, "aCol", DecimalType.of(20, 1)));
+    Schema newSchema = new Schema(required(1, "aCol", DecimalType.of(22, 1)));
+
+    Schema applied = new SchemaUpdate(currentSchema, 1).upsertSchema(newSchema).apply();
+    Assert.assertEquals(newSchema.asStruct(), applied.asStruct());
+  }
+
+  @Test
+  public void testAddPrimitiveToNestedStruct() {
+    Schema schema = new Schema(
+        required(1, "struct1", Types.StructType.of(
+          optional(2, "struct2", Types.StructType.of(
+                  optional(3, "list", Types.ListType.ofOptional(
+                      4, Types.StructType.of(
+                          optional(5, "value", StringType.get())))))))));
+
+    Schema newSchema = new Schema(
+        required(1, "struct1", Types.StructType.of(
+            optional(2, "struct2", Types.StructType.of(
+                optional(3, "list", Types.ListType.ofOptional(
+                    4, Types.StructType.of(
+                        optional(5, "time", TimeType.get())))))))));
```

Comment:
seems like "value" is missing from newSchema.  I think deletes are not handled that's why. We should consider adding deletes and renames too

Reply:
Right, nor should it, the main use-case for this API is to accommodate schema unions as @rdblue's summed it up during our chat  in the Iceberg sync-up today.
The use-case is as follows:

1.  think a dataset with a large schema - call it `effective schema` 
2. we get data that only contains particular columns - those columns make up only for a subset of the large schema - call it `observed schema` 
3. as these chunks of data and their respective observed schemas are being landed in the iceberg table we want to keep the table schema on-par with the union of these observed schemas

So in this particular `expected` is the union of `schema` and `newSchema`. 
Does this make sense? 
Does this resemble your use-case as well?

Reply:
Yes, that use case makes sense to me! It will work for our use case too as we do not support deletes

Reply:
What about renames?

Reply:
Dunno really, do you need renames in your use-case?
If so, do you have a plan on how to push those changes down to Iceberg APIs, do you know before committing to Iceberg what the effective columns you're renaming are? Would we rely on Iceberg to perform the renames implicitly, how would it differentiate between an add or a rename?
 
In the particular use-case I've described I didn't account for that requirement. Observed schemas don't require column renames.
However I assume that the only way we could reliably implement this is to reconcile column name changes by relying on the field ids though, right? 
Also for this particular use-case of sub-schemas getting unioned to support column renames the operations are additive so they can be running out of order - eventually the union schema is the same no matter the order in which the observed schemas were committed. 
But for things like renames and deletes we'd need to account for order too, right?
I hate building things that need to account for order cause most of the time I get them wrong.
So I'd say that I don't think that we can combine the two implementations into a single method...
Wdyt?

Reply:
Yea, @rdblue had some thoughts on renames as well https://github.com/apache/iceberg/pull/759#discussion_r382871744 . I'm not sure if that thinking has evolved in the last meetup.

Reply:
I'll find time to look at this tomorrow, hopefully. Sorry for the delay.

Reply:
I just went back to the sync notes to refresh my memory and found that I hadn't posted them to the doc! I updated it. Sorry about that.

Just catching up on this now. I think the main use case that we're trying to solve here is the union case, where you just need to produce a schema with a union of the fields from all of the observed schemas. I think focusing on that use case makes the choice of whether to support delete clear: just because a column is missing, doesn't mean it shouldn't be in the union. Renames are similar: if the field exists, do nothing and if it doesn't exist, add it.

The only case where we could detect a rename in a union operation is if both schemas have IDs. But then I would say that we should avoid unnecessary modification of the union schema. That's because we don't know which name is correct (the existing name or the observed name) but using the observed name is more likely to break existing queries for the table.